DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -119,23 +73,67 @@ internal class DataUploadWorker: DataUploadWorkerType {
+                } else {
+                    self.fileReader.markBatchAsRead(batch, reason: .intakeCode(responseCode: uploadStatus.responseCode ?? -1)) // -1 is unexpected here
+                    self.delay.decrease()
+
+                    DD.logger.debug("   → (\(self.featureName)) accepted, won't be retransmitted: \(uploadStatus.userDebugDescription)")
+                }
+
+                switch uploadStatus.error {
+                case .unauthorized:
+                    DD.logger.error("⚠️ Make sure that the provided token still exists and you're targeting the relevant Datadog site.")
+                case let .httpError(statusCode: statusCode):
+                    DD.telemetry.error("Data upload finished with status code: \(statusCode)")
+                case let .networkError(error: error):
+                    DD.telemetry.error("Data upload finished with error", error: error)
+                case .none: break
+                }
+            } catch let error {
+                // If upload can't be initiated do not retry, so drop the batch:
+                self.fileReader.markBatchAsRead(batch, reason: .invalid)
+                DD.telemetry.error("Failed to initiate '\(self.featureName)' data upload", error: error)
+            }
+        } else {
+            let batchLabel = nextBatch != nil ? "YES" : (isSystemReady ? "NO" : "NOT CHECKED")
+            DD.logger.debug("💡 (\(self.featureName)) No upload. Batch to upload: \(batchLabel), System conditions: \(blockersForUpload.description)")
+
+            self.delay.increase()
+        }
+        if let taskID = taskID {
+            self.backgroundTaskCoordinator?.endBackgroundTaskIfActive(taskID)
+        }
+    }
+
     /// Sends all unsent data synchronously.
     /// - It performs arbitrary upload (without checking upload condition and without re-transmitting failed uploads).
     internal func flushSynchronously() {
-        queue.sync {
-            while let nextBatch = self.fileReader.readNextBatch() {
-                defer {
-                    // RUMM-3459 Delete the underlying batch with `.flushed` reason that will be ignored in reported
-                    // metrics or telemetry. This is legitimate as long as `flush()` routine is only available for testing
-                    // purposes and never run in production apps.
-                    self.fileReader.markBatchAsRead(nextBatch, reason: .flushed)
-                }
-                do {
-                    // Try uploading the batch and do one more retry on failure.
-                    _ = try self.dataUploader.upload(events: nextBatch.events, context: contextProvider.read())
-                } catch {
-                    _ = try? self.dataUploader.upload(events: nextBatch.events, context: contextProvider.read())
-                }
+        queue.sync { [weak self] in
+            while let nextBatch = self?.fileReader.readNextBatch() {
+                self?.tryToSendBatch(nextBatch)
```

Comment:
The new `tryToSendBatch()` call involves the full retry logic. This means: if calling `flushSynchronously()` under insufficient upload conditions (e.g. offline), the caller thread will run into infinite while loop.

See `flushSynchronously()` comment:
> /// - It performs arbitrary upload (without checking upload condition and without re-transmitting failed uploads).